sql,catalog/descs,gcjob,lease,importccl,backupccl: replace PublishMultiple

ajwerner opened the pull request:
This series of commits removed `lease.Manager.PublishMultiple()` and its variants and replaces them with a new transactional wrapper around the `descs.Collection`.

Fixes #50906
Fixes #51417

This is WIP on the road to giving a similar treatment to restore and import as a part of #53378.

Commit:
sql: add commentary around leasing of descriptors during backfill

Release justification: non-production code changes

Release note: None

Commit:
sql: clean up backfill internal planner cleanup

Bit of a drive-by.

Release justification: bug fixes and low-risk updates to new functionality

Release note: None

Commit:
sql,catalog/descs: extract CheckTwoVersionInvariant, add descs.Txn

Release justification: low risk, high benefit changes to existing functionality
Release note: None

Commit:
sql: adopt descs.Txn in type and schema changer

This commit removes all uses of `lease.Publish` and `lease.PublishMultiple`

Release justification: low risk, high benefit change to existing functionality.

Release note: None

Commit:
catalogkv: extract a function to write namespace entry removal to a batch

Release justification: low-risk updates to new functionality

Release note: None

Commit:
sql,catalog/descs: add descs.Collection.WriteDescToBatch

This method is handy as all over the place we want to consider incrementing the
version, add the uncommitted descriptor to the `Collection` and then write it
to a batch. We use this more later.

It also permits adding an uncommitted descriptor which is new. This enables
adding a new descriptor that has a version other than 1.

Release justification: low-risk updates to new functionality.

Release note: None

Commit:
catalog/descs: export GetMutableDescriptorByID

Release justification: low-risk updates to new functionality

Release note: None

Commit:
gcjob: replace lease.Manager.Publish with descs.Txn

Release justification: low risk, high benefit changes to existing functionality

Release note: None

Commit:
catalog/lease: move Publish and PublishMultiple to testing

Release justification: non-production code changes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53727)
<!-- Reviewable:end -->

Commit:
importccl,backupccl: adopt descs.Txn and general descriptor handling

This commit adopts descs.Txn and the relevant new descs.Collection methods in
restore. It should eliminate all of the cases where we might have written a
mutated descriptor that was not read from the store. It also eliminates the
conditional reading of descriptors which in the later step was deemed to be
somewhat paranoid. In the initial transaction of the two jobs it makes a
version mismatch into an error.

Release justification: low risk, high benefit changes to existing functionality

Release note: None

Commit:
tabledesc: don't increment the version in MaybeIncrementVersion with Created

This causes problems. In the other descriptors we use a pointer for the
cluster version and then check on that nil in `MaybeIncrementVersion`. For
legacy reasons, it is not a pointer in the `tabledesc.Mutable`. Nevertheless,
make the logic match.

Release justification: bug fixes and low-risk updates to new functionality

Release note: None

ajwerner commented:
@lucy-zhang I would still like to add more unit testing to this but I think it is ready for review. I'll admit that it touches more than is maybe ideal but I think it leaves us in a happier place. LMKWYT

ajwerner commented:
@pbardea it'd be wonderful if you were willing to take a look at the last commit in particular. It's in many ways the meat of this change.

ajwerner commented:
Okay I've rebased on top of #53848

pbardea reviewed (commented):
Did a first pass on the last commit and it looks good to me
Thanks for rebasing; sorry for not CC'ing you earlier.

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 5 of 5 files at r6, 2 of 2 files at r7, 1 of 5 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-:-MGK4i2e9lokQkwMCwpz:b-curooq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 810 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGK8S9p3i7v-H1zoqoo:-MGK8S9p3i7v-H1zoqop:b-sftw98) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2c5dbf663ce58b4c119ed36e370a330fa95749/pkg/ccl/backupccl/restore_job.go#L810)):*
> ```Go
> 		case catalog.TableDescriptor:
> 			mut := tabledesc.NewCreatedMutable(*desc.TableDesc())
> 			mut.Version = 1
> ```

I think it's interesting that we need to reset the version of ingested descriptors. Is there a comment that explains why this is needed? Maybe a short note around here would be helpful.
___
*[pkg/ccl/backupccl/restore_job.go, line 1355 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGK4cZL39HT6lbx9P5S:-MGK4cZMA0jfTRfFy4p2:btx2v11) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2c5dbf663ce58b4c119ed36e370a330fa95749/pkg/ccl/backupccl/restore_job.go#L1355)):*
> ```Go
> 		// it's not really clear what to do. Just log and carry on.
> 		if got, exp := mutableTables[i].Version, details.TableDescs[i].Version; got != exp {
> 			log.Errorf(ctx, "version changed for restored descriptor before "+
> ```

Let's also log the descriptor ID here?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 810 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGK8S9p3i7v-H1zoqoo:-MGKAoWg70d2RXyjGk7F:bwmhi28) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2c5dbf663ce58b4c119ed36e370a330fa95749/pkg/ccl/backupccl/restore_job.go#L810)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

I think it's interesting that we need to reset the version of ingested descriptors. Is there a comment that explains why this is needed? Maybe a short note around here would be helpful.
</blockquote></details>

I made it so we don't need to do this. I'll leave it to Lucy to do this in https://github.com/cockroachdb/cockroach/pull/53539.
___
*[pkg/ccl/backupccl/restore_job.go, line 1355 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGK4cZL39HT6lbx9P5S:-MGKCLw88nfwjGEkB5jY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc2c5dbf663ce58b4c119ed36e370a330fa95749/pkg/ccl/backupccl/restore_job.go#L1355)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Let's also log the descriptor ID here?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

lucy-zhang commented on pkg/ccl/importccl/import_stmt.go:
```diff
@@ -1319,15 +1316,16 @@ func (r *importResumer) publishTables(ctx context.Context, execCfg *sql.Executor
 	log.Event(ctx, "making tables live")
 
 	// Needed to trigger the schema change manager.
```
nit: this comment is very outdated

lucy-zhang commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -1350,21 +1335,28 @@ func (r *restoreResumer) dropDescriptors(
+		if err != nil {
+			return err
+		}
+		// Ensure that the version matches we expect. In the case that it doesn't,
```
nit: `what we expect`

lucy-zhang replied:
IIUC the descriptors could either still be offline or be public at this point. If they're already public, we can't actually make any guarantees about their versions, right?

lucy-zhang commented on pkg/sql/catalog/descs/txn.go:
```diff
@@ -0,0 +1,184 @@
+	leaseMgr *lease.Manager,
+	ie sqlutil.InternalExecutor,
+	db *kv.DB,
+	f func(ctx context.Context, txn *kv.Txn, descriptors *Collection) error,
```
nit: I might prefer this argument be `descCollection` or something. `descriptors` always makes me think we're calling functions in a `descriptors` package. I don't feel strongly about this at all.

thoszhang reviewed (commented):
I think you squashed two commits into your last commit and didn't update the default message.

Commit:
catalog/descs: deal somewhat with offline states resolving user-defined types

This commit ensures that offline schemas can be resolved in the context of
resolving mutable table descriptors which utilize user-defined types.

Release note: None

ajwerner reviewed (commented):
Rebased and RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-:-MHB9Pwu0iBZj5Iob0G_:b-1ti1ar)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 1348 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGJzq6jDufUWZ1L06Jd-r30-1348:-MHAfSSV8BU3IGXPnTTD:bspifj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/16a5cda799a177750573dab46dd9864d084e1e05/pkg/ccl/backupccl/restore_job.go#L1348)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

IIUC the descriptors could either still be offline or be public at this point. If they're already public, we can't actually make any guarantees about their versions, right?
</blockquote></details>

Done. Only check now if we haven't published the descriptors.
___
*[pkg/sql/schema_changer.go, line 904 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MG7IuU9UdrgdzqPtcdt-r27-904:-MHAWOim9HEnQlEYPFSY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfdc19333c229f625f9d621c93c680514f9e6932/pkg/sql/schema_changer.go#L904)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: update this comment
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/txn.go, line 45 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MG7IuU9UdrgdzqPtce0-r26-45:-MHAWlzY3E77xlkQp9yi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/374364eda1cb3dbb2707a397bbec9b111d731105/pkg/sql/catalog/descs/txn.go#L45)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: I might prefer this argument be `descCollection` or something. `descriptors` always makes me think we're calling functions in a `descriptors` package. I don't feel strongly about this at all.
</blockquote></details>

Done.
___
*[pkg/sql/catalog/lease/helpers_test.go, line 123 at r29](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MGJzq6jDufUWZ1L06Jb-r29-123:-MHAZh0s4XPPCzGEmVW1:b-dkyl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8237ac6d18d26fe7af3794b33182e6cdd340ff4c/pkg/sql/catalog/lease/helpers_test.go#L123)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

We can get rid of this now, right? It doesn't even seem to be used in the tests besides a test for the `ErrDidntUpdateDescriptor` functionality itself.
</blockquote></details>

sure, done, I'd love to kill Publish and PublishMultiple too but that can be follow-up. At least they now live in testing.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
> Reviewed 1 of 2 files at r21, 2 of 2 files at r22, 4 of 5 files at r25, 2 of 2 files at r26, 5 of 5 files at r27, 2 of 2 files at r28, 2 of 2 files at r29, 4 of 4 files at r30, 1 of 2 files at r32, 2 of 2 files at r33, 5 of 7 files at r36, 2 of 2 files at r37, 5 of 5 files at r38, 2 of 2 files at r39, 3 of 6 files at r40, 5 of 5 files at r41, 2 of 2 files at r42.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 999 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHBko7a3Gn3H7H7zuXC:-MHBko7a3Gn3H7H7zuXD:br25gss) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L999)):*
> ```Go
> 						}
> 						typDesc.AddReferencingDescriptorID(table.GetID())
> 						// TODO (lucy): I think we should be incrementing the version for the
> ```

We can get rid of this comment now.
___
*[pkg/ccl/backupccl/restore_job.go, line 1262 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHC71QO9vpJ-ApHAOCO:-MHC71QP9K5rvsmNZ7bN:ba62m03) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1262)):*
> ```Go
> func (r *restoreResumer) publishDescriptors(
> 	ctx context.Context, txn *kv.Txn, descsCol *descs.Collection,
> ) (newDescriptorChangeJobs []*jobs.StartableJob, err error) {
> ```

Returning the jobs is kind of weird because the jobs just wait for leases to drain, but we don't actually wait for the jobs to finish, right? I'm fine with this as-is since I think it's good consistency-wise for schema changes to always be tied to jobs, but the extra complexity for no good reason is kind of annoying.
___
*[pkg/ccl/backupccl/restore_job.go, line 1387 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHBmDbJ0NQBz5S_AAlo:-MHBmDbJ0NQBz5S_AAlp:b-fqsjnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1387)):*
> ```Go
> 
> 	if err := txn.Run(ctx, b); err != nil {
> 		return newDescriptorChangeJobs, errors.Wrap(err, "publishing tables")
> ```

Is this intentional?
___
*[pkg/sql/catalog/descs/txn.go, line 45 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MG7IuU9UdrgdzqPtce0-r26-45:-MHBpE8i-k0XVxa3gjG2:bbs25uq) ([raw file](https://github.com/cockroachdb/cockroach/blob/374364eda1cb3dbb2707a397bbec9b111d731105/pkg/sql/catalog/descs/txn.go#L45)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Could you also update this in all the other files?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 999 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHBko7a3Gn3H7H7zuXC:-MHCVlU4DdJEE20BO3hI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L999)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

We can get rid of this comment now.
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/restore_job.go, line 1262 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHC71QO9vpJ-ApHAOCO:-MHCWF6i1btUw_Kb2UFL:b-c8nnga) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1262)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Returning the jobs is kind of weird because the jobs just wait for leases to drain, but we don't actually wait for the jobs to finish, right? I'm fine with this as-is since I think it's good consistency-wise for schema changes to always be tied to jobs, but the extra complexity for no good reason is kind of annoying.
</blockquote></details>

it's important because you can't start them until the transaction commits and the transaction is handled above this call. Added a comment. 
___
*[pkg/ccl/backupccl/restore_job.go, line 1387 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHBmDbJ0NQBz5S_AAlo:-MHCWAUF1Y0JHmWKzila:biuwarn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1387)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Is this intentional?
</blockquote></details>

this being? returning the jobs, yes, there's a defer to clean them up. 
___
*[pkg/sql/catalog/descs/txn.go, line 45 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MG7IuU9UdrgdzqPtce0-r26-45:-MHCWNt50WrhNfla7Rcu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/374364eda1cb3dbb2707a397bbec9b111d731105/pkg/sql/catalog/descs/txn.go#L45)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Could you also update this in all the other files?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
> Reviewed 1 of 16 files at r43.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53727)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @lucy-zhang, and @pbardea)

___
*[pkg/ccl/backupccl/restore_job.go, line 1262 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHC71QO9vpJ-ApHAOCO:-MHCZ7WeD_pmFmL53eRJ:b-lewgce) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1262)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

it's important because you can't start them until the transaction commits and the transaction is handled above this call. Added a comment. 
</blockquote></details>

Sorry, my point wasn't about returning the jobs per se, it was more that we even start the jobs at all when the jobs do nothing except wait for leases to drain, but we don't wait for the jobs to finish within this job. I'm fine with this as-is, but it's just unsatisfying.
___
*[pkg/ccl/backupccl/restore_job.go, line 1387 at r42](https://reviewable.io/reviews/cockroachdb/cockroach/53727#-MHBmDbJ0NQBz5S_AAlo:-MHCZdXC7owtaa5P_Vqs:b2xiksc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b9916c50a4f57e4df413ba9d0ddf697500dd65c/pkg/ccl/backupccl/restore_job.go#L1387)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this being? returning the jobs, yes, there's a defer to clean them up. 
</blockquote></details>

Sorry, I was looking at an earlier revision, never mind.


<!-- Sent from Reviewable.io -->

ajwerner commented:
TFTR!

bors r+
